sympy/functions/elementary/piecewise.py

```diff
@@ -664,3 +967,222 @@ def piecewise_fold(expr):
             new_args.append((expr.func(*e), And(*c)))
 
     return Piecewise(*new_args)
+
+
+def _pair(A, B, union=False, dict=None, abi=False):
+    """Return the intervals on the real number line that are covered
+    by the two intervals, A = [a, b] and B = [c, d]. If union is
+    true then a simple list of intervals is returned; if dict is
+    true (default) then the intervals are keyed to 0 if they are
+    covered by A, 1 if they are covered by B and -1 if they are covered
+    by neither, with higher precedence given to A. If abi is true,
+    return an ordered set of tuples (a, b, i) where
+    a is the lower bound, b is the upper bound and i indicates which
+    interval covers that range: 0 for A, 1 for B and -1 for neither.
+
+    Examples
+    ========
+
+    >>> from sympy import Tuple
+    >>> from sympy.functions.elementary.piecewise import _pair
+    >>> from sympy.abc import x
+
+    For the following segments,
+
+           1---2
+                      4---5
+
+    5 regions are identified:
+
+        -oo -> 1 -> 2 -> 4 -> 5 -> oo
+
+    >>> A = (1, 2); B = (4, 5)
+    >>> _pair(A, B, abi=True)
+    [(-oo, 1, -1), (1, 2, 0), (2, 4, -1), (4, 5, 1), (5, oo, -1)]
+    >>> _pair(A, B, union=True)
+    [(1, 2), (4, 5)]
+    >>> _pair(A, B, dict=True)  # default
+    {-1: [(-oo, 1), (2, 4), (5, oo)], 0: [(1, 2)], 1: [(4, 5)]}
+
+    When one or more of the boundaries are symbolic, the results
+    become more complex and the boundaries are given in terms of
+    Min and Max functions:
+
+    >>> _pair((x, 2), (4, 5))
+    {-1: [(-oo, Min(4, x)), (Min(5, x), x), (2, 4), (5, oo)],
```

Comment:
Couldn't this be simpler, since x has to be <= 2?

Reply:
I believe this is not simpler because it aims to be right even if x is greater than 2. If you follow along in the docstring, different values are given for x. I'm not sure of the implications if the assumption is made that x <= 2. It's probably worth revisiting. I'll take a look. One might even be able to trick the mins into resolving by supplying a known displacement to one arg, e.g. if x is assumed to be less than 2 then x could be replaced with `2 - nonnegative` and then `Min(2 - nonnegative, 4)` would resolve to `2 - nonnegative` and backsubstituting nonnegative with `2 - x` would yield x again instead of `Min(x, 4)`.

Reply:
I'm working on the assumption that if x >= 2 then the interval (x, 2) is empty. In that case you have just -oo -- 4 -- 5 -- oo. I would need to think about this. It's possible I am wrong that it can be simplified. Certainly in a case like (2, x), (4, 5) it cannot be simplified.